teps/0103-skipping-reason.md

```diff
@@ -0,0 +1,215 @@
+---
+status: implementable
+title: Skipping Reason
+creation-date: '2022-04-06'
+last-updated: '2022-04-06'
+authors:
+- '@jerop'
+see-also:
+- TEP-0007
+- TEP-0059
+- TEP-0100
+---
+
+# TEP-0103: Skipping Reason
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Example](#example)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Today, users only know that a `PipelineTask` was skipped, but they don't know for which
+exact reason. This can be confusing when debugging `Pipelines`. In this TEP, we propose
+adding the reason for skipping to the `SkippedTasks` field in `PipelineRunStatus` to
+improve usability and debuggability.
+
+## Motivation
+
+There are many [reasons][reasons] why a `PipelineTask` could be skipped, including:
+* at least one of its `when` expressions evaluated to false 
+* at least one of its `Conditions` failed
+* at least one of the `Results` it was consuming was missing
+* at least one its parent `PipelineTasks` was skipped
+* the `PipelineRun` was in stopping state
+* the `PipelineRun` was gracefully cancelled
+* the `PipelineRun` was gracefully stopped
+
+When users see a `PipelineTask` has been skipped without knowing the reason, they may get
+confused about the behavior of the `PipelineRun` e.g. [issue][issue] and [thread][slack].
+
+### Goals
+
+* Surface the reason why a specific `Task` in a `Pipeline` was skipped.
+
+### Use Cases
+
+* As a `Pipeline` user, I need to know why certain `Tasks` in my `Pipeline` have been skipped.
+* As a maintainer of *Tekton Pipelines*, I need to test that certain `Tasks` in a `Pipeline`
+  are skipped for the exact reason I expect them to be skipped for.
+
+### Requirements
+
+* Users can find out the reason why a given `Task` in a `Pipeline` was skipped.
+
+## Proposal
+
+To surface the reason for which a `PipelineTask` was skipped, we propose adding `"reason"` field
+to the [`SkippedTasks`][skipped-tasks] field in `PipelineRunStatus`.
+
+```go
+// SkippedTask is used to describe the Tasks that were skipped due to their When Expressions
+// evaluating to False. This is a struct because we are looking into including more details
+// about the When Expressions that caused this Task to be skipped.
+
+type SkippedTask struct {
+	
+	// Name is the Pipeline Task name
+	Name string `json:"name"`
+	
+	// Reason is the cause of the PipelineTask being skipped
+	Reason SkippingReason `json:"reason"`
```

Comment:
any reason not to put this info into TaskRun.Status.Condition.Reason instead of creating a separate field? what is in the Condition.Reason when a TaskRun is skipped?

Reply:
> any reason not to put this info into TaskRun.Status.Condition.Reason instead of creating a separate field? what is in the Condition.Reason when a TaskRun is skipped?

We do not create a `TaskRun` for a skipped `PipelineTask`, therefore there's no `TaskRun.Status.Condition.Reason` to put this information -- this is why we added the `SkippedTasks` field to the `PipelineRun`

Reply:
ah whoops 😬